Embrace Unity iOS Interface/EmbraceManager.swift

```diff
@@ -0,0 +1,488 @@
+            "emb.type": "sys.exception",
+            "exception.message": message,
+            "exception.type": name
+        ]
+        
+        Embrace.client?.log("Unity exception",
+                            severity: .error,
+                            timestamp: Date(), // Should we let users input their own exception timestamp? That would create an API mismatch among the Hosted SDKs. Let's not.
+                            attributes: attributes)
+    }
+    
+    static func logPushNotification(title: String, body: String, subtitle: String, badge: Int, category: String) -> Bool {
+        
+        // Borrowed from Embrace Flutter SDK
+        let pushData: [AnyHashable: Any?] = [
+                        "aps": [
+                            "alert" : [
+                                "title" : title,
+                                "subtitle" : subtitle,
+                                "body" : body
+                            ],
+                            "badge" : badge as Any,
+                            "category" : category as Any
+                        ]
+                    ]
+        do {
+            try Embrace.client?.add(event: .push(userInfo: pushData as [AnyHashable: Any]))
+            return true
+        } catch let error {
+            // We should probably log this error
+        }
+        
+        return false
+    }
+    
+    private static func transferKVPs(dest: inout [String:String], src: NSDictionary) {
+        for (key, value) in src {
+            if let key = key as? String, let value = value as? String {
+                dest.updateValue(value, forKey: key)
+            }
+        }
+    }
+    
+    private static func convertNSDictToSwiftDict<T>(nsDict: NSDictionary, converter: (String) -> T) -> [String: T] {
+        var swiftDict = [String: T]()
+        
+        for (key, value) in nsDict {
+            if let key = key as? String, let value = value as? String {
+                swiftDict.updateValue(converter(value), forKey: key)
+            } else {
+                // Should the whole operation fail or should we just log the issue?
+                // We probably should not let the whole operation fail. We can skip the value
+            }
+        }
+        
+        return swiftDict
+    }
+    
+    private static func createNetworkSpanName(url: String, httpMethod: String) -> String {
+        var name = "emb-" + httpMethod.uppercased()
+        
+        if let fullUrl = URL(string: url) {
+            let path = fullUrl.path
+            if (!path.isEmpty && path != "/") {
+                name += " " + path
+            }
+        }
+        
+        return name
+    }
+    
+    // Assumes values in Unix Epoch time
+    private static func convertDoubleToDate(ms: Double) -> Date {
+        return Date(timeIntervalSince1970: TimeInterval(ms / 1000.0))
+    }
+    
+    private static func convertStringToErrorCode(str: String) -> ErrorCode? {
```

Comment:
I might be missing a step but I had seen in `io.embrace.sdk/Scripts/Native/Embrace_iOS.cs` that error code was being passed through as an int, does that get mapped to a string at some point before hitting this?

Reply:
It does. It's actually kind of silly looking at the whole chain.

The Embrace Unity SDK API expects an `EmbraceSpanErrorCode` enum on the public facing API. Then it internally converts that to an `int?`. Which then gets converted to a C# `string` before being passed over to the Swift side as a c-string and converted to the associated Swift enum there. I'm not *entirely* sure why this is what happened in the first place.

Externally I'm pretty sure we can't update the C# external facing API, but internally we could probably smooth that over. I know we want to go to Unity 2.0 at some point soon in the future and cleaning this up would be 100% part of my goal.